File: services/userService.js
// /services/userService.js

const { v4: uuidv4 } = require("uuid");
const bcrypt = require("bcrypt");
const db = require("../db"); // presupposto: layer DB astratto

class UserService {
  async createUser({ email, password, name, role }) {
    const hashedPassword = await bcrypt.hash(password, 12);
    const user = {
      id: uuidv4(),
      email,
      password: hashedPassword,
      name,
      role: role || "user",
      createdAt: new Date(),
      updatedAt: new Date(),
    };
    await db.insert("users", user);
    return user;
  }

  async getUserByEmail(email) {
    return db.findOne("users", { email });
  }

  async getUserById(id) {
    return db.findOne("users", { id });
  }

  async updateUser(id, updates) {
    updates.updatedAt = new Date();
    await db.update("users", id, updates);
    return this.getUserById(id);
  }

  async deleteUser(id) {
    await db.delete("users", id);
  }

  async verifyPassword(email, plainPassword) {
    const user = await this.getUserByEmail(email);
    if (!user) return false;
    return bcrypt.compare(plainPassword, user.password);
  }
}

module.exports = new UserService();

File: services/user_service.py
"""
user_service.py – Gestione utenti: CRUD, onboarding, sicurezza, ruoli.
"""
from models.user import User
import uuid
from datetime import datetime

class UserService:
    def __init__(self, db):
        self.db = db

    def create_user(self, email: str, name: str, password_hash: str, lang="it", role="user", meta=None):
        user = User(
            id=str(uuid.uuid4()),
            email=email,
            name=name,
            password_hash=password_hash,
            role=role,
            lang=lang,
            created_at=datetime.utcnow(),
            is_active=True,
            meta=meta or {}
        )
        self.db.save(user)
        return user

    def get_user(self, user_id: str):
        return self.db.get(User, user_id)

    def update_user(self, user_id: str, **kwargs):
        user = self.get_user(user_id)
        if not user:
            return None
        for k, v in kwargs.items():
            setattr(user, k, v)
        self.db.save(user)
        return user

    def delete_user(self, user_id: str):
        user = self.get_user(user_id)
        if user:
            user.is_active = False
            self.db.save(user)
        return user
